samples/cpp/macbeth_chart_detection.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+
+    keys += genPreprocArguments(modelName, zooFile);
+    parser = CommandLineParser(argc, argv, keys);
+
+    int t = parser.get<int>("type");
+
+    CV_Assert(0 <= t && t <= 2);
+    COLORCHART chartType = COLORCHART(t);
+
+    const string sha1 = parser.get<String>("sha1");
+    const string model_path = findModel(parser.get<string>("model"), sha1);
+    const string config_sha1 = parser.get<String>("config_sha1");
+    const string pbtxt_path = findModel(parser.get<string>("config"), config_sha1);
+    const string backend = parser.get<String>("backend");
+    const string target = parser.get<String>("target");
+
+    int nc = parser.get<int>("num_charts");
+
+    Ptr<CCheckerDetector> detector;
+    if (model_path != "" && pbtxt_path != ""){
+        EngineType engine = ENGINE_AUTO;
+        if (backend != "default" || target != "cpu"){
+            engine = ENGINE_CLASSIC;
+        }
+        Net net = readNetFromTensorflow(model_path, pbtxt_path, engine);
+        net.setPreferableBackend(getBackendID(backend));
+        net.setPreferableTarget(getTargetID(target));
+
+        detector = CCheckerDetector::create(net);
+        cout<<"Detecting checkers using neural network."<<endl;
+    }
+    else{
+        detector = CCheckerDetector::create();
+    }
+
+    bool isVideo = true;
+    Mat image;
+    VideoCapture cap;
+
+    if (parser.has("input")){
+        const string inputFile = parser.get<String>("input");
+        image = imread(findFile(inputFile));
+        if (!image.empty())
+        {
+            isVideo = false;
+        }
+        else
+        {
+            // Not an image, so try opening it as a video.
+            cap.open(findFile(inputFile));
+            if (!cap.isOpened())
+            {
+                cout << "[ERROR] Could not open file as an image or video: " << inputFile << endl;
+                return -1;
+            }
+        }
+    }
+    else
+        cap.open(0);
+
+    Mat src, tgt;
+    if (isVideo){
+        cout<<"To print the actual colors and reference colors for current frame press SPACEBAR. To resume press SPACEBAR again"<<endl;
+
+        while (cap.grab())
+        {
+            Mat frame;
+            cap.retrieve(frame);
+
+            processFrame(frame, detector, chartType, src, tgt, nc);
+
+            int key = waitKey(10);
+            if (key == ' '){
+                cout<<"Reference colors: "<<tgt<<endl<<"--------------------"<<endl;
+                cout<<"Actual colors: "<<src<<endl<<endl;
+                cout<<"Press spacebar to resume."<<endl;
+
+                int pauseKey = waitKey(0);
+                if (pauseKey == 27) {
+                    exit(0);
+                }
+                cout << "Resumed! Processing continues..." << endl;
+            }
+            else if (key == 27) exit(0);
```

Comment:
1) when user presses ESCAPE, he does not get colors printed.
2) what if the board was never detected?